gpii/node_modules/userListeners/src/userFolder.js

```diff
@@ -0,0 +1,115 @@
+/* userFolder user listener.
+ *
+ * Copyright 2017 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * The R&D leading to these results received funding from the
+ * Department of Education - Grant H421A150005 (GPII-APCP). However,
+ * these results do not necessarily represent the policy of the
+ * Department of Education, and you should not assume endorsement by the
+ * Federal Government.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fluid = require("infusion");
+
+var gpii = fluid.registerNamespace("gpii");
+
+// The user folder user listener.
+fluid.defaults("gpii.userListeners.userFolder", {
+    gradeNames: ["fluid.component", "fluid.contextAware", "gpii.userListener"],
+    contextAwareness: {
+        platform: {
+            checks: {
+                windows: {
+                    contextValue: "{gpii.contexts.windows}",
+                    gradeNames: "gpii.userListeners.userFolder.windows"
+                }
+            }
+        }
+    },
+    members: {
+        tokenDirectory: "@expand:{settingsDir}.getGpiiSettingsDir()",
+        tokenFile: ".gpii-user-token.txt",
+        done: false,
+        listenerName: "userFolder"
+    },
+    components: {
+        settingsDir: {
+            type: "gpii.settingsDir"
+        }
+    },
+    invokers: {
+        startListener: "{that}.attemptLogin",
+        stopListener: "fluid.identity",
+        attemptLogin: {
+            funcName: "gpii.userListeners.userFolder.attemptLogin",
+            args: ["{that}"]
+        },
+        getToken: {
+            func: "{that}.readTokenFile",
+            args: ["{that}.tokenDirectory", "{that}.tokenFile"]
+        },
+        storeToken: {
+            func: "{that}.writeTokenFile",
+            args: [ "{that}.tokenDirectory", "{that}.tokenFile", "{arguments}.0" ]
+        }
+    },
+    listeners: {
+        "onCreate.loginListener": {
+            funcName: "gpii.userListeners.userFolder.addLoginListener",
+            args: [ "{that}", "{lifecycleManager}.events.onSessionStart"]
+        }
+    },
+    // Disabling by default, as it is the safest and will prevent people's keys from being accidentally stored when it
+    // is not desired (eg, on shared or public access computers).
+    saveLastLogin: false
+});
+
+/**
+ * Adds a listener to the onSessionStart event of the lifecycle manager, which stores the token file.
+ *
+ * @param {Component} that The gpii.userListeners.userFolder instance.
+ * @param {EventFirer} firer The event firer for the onSessionStart event.
+ */
+gpii.userListeners.userFolder.addLoginListener = function (that, firer) {
+    if (that.options.saveLastLogin) {
+        firer.addListener(function (gradeName, gpiiKey) {
+            if (gpiiKey !== "noUser") {
```

Comment:
You might like to check the gpiiKey is not any of reserved keys. A list of reserved keys can be found [here|https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/SystemUtils.js#L19|.